engines/e_pkcs11.c

```diff
@@ -0,0 +1,1026 @@
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjectsFinal(session);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsFinal failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_FIND_PUBLIC_KEY,
+                  PKCS11_R_FIND_OBJECT_FINAL_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_GetAttributeValue(session, key,
+                                           tmpl, OSSL_NELEM(tmpl));
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetAttributeValue failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_FIND_PUBLIC_KEY,
+                  PKCS11_R_GETATTRIBUTEVALUE_FAILED);
+        goto err;
+    }
+    return key;
+
+ err:
+    return 0;
+}
+
+EVP_PKEY *pkcs11_load_pkey(CK_SESSION_HANDLE session, PKCS11_CTX *ctx,
+                           CK_OBJECT_HANDLE key)
+{
+    EVP_PKEY *k = NULL;
+    CK_RV rv;
+    CK_ATTRIBUTE rsa_attributes[2];
+    RSA *rsa = NULL;
+
+    rsa_attributes[0].type = CKA_MODULUS;
+    rsa_attributes[0].pValue = NULL;
+    rsa_attributes[0].ulValueLen = 0;
+    rsa_attributes[1].type = CKA_PUBLIC_EXPONENT;
+    rsa_attributes[1].pValue = NULL;
+    rsa_attributes[1].ulValueLen = 0;
+    rsa = RSA_new();
+
+    if (rsa == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_LOAD_PKEY, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_GetAttributeValue(session, key,
+                                           rsa_attributes,
+                                           OSSL_NELEM(rsa_attributes));
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetAttributeValue failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_LOAD_PKEY,
+                  PKCS11_R_GETATTRIBUTEVALUE_FAILED);
+        goto err;
+    }
+    if  (rsa_attributes[0].ulValueLen == 0
+         || rsa_attributes[1].ulValueLen == 0)
+        goto err;
+
+    rsa_attributes[0].pValue = OPENSSL_malloc(rsa_attributes[0].ulValueLen);
+    if (rsa_attributes[0].pValue == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_LOAD_PKEY, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    rsa_attributes[1].pValue = OPENSSL_malloc(rsa_attributes[1].ulValueLen);
+    if (rsa_attributes[1].pValue == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_LOAD_PKEY, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_GetAttributeValue(session, key,
+                                           rsa_attributes,
+                                           OSSL_NELEM(rsa_attributes));
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_GetAttributeValue failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_LOAD_PKEY,
+                  PKCS11_R_GETATTRIBUTEVALUE_FAILED);
+        goto err;
+    }
+
+    k = EVP_PKEY_new();
+    if (k == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_LOAD_PKEY, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    RSA_set_ex_data(rsa, rsa_pkcs11_idx, (char *) (CK_OBJECT_HANDLE) key);
```

Comment:
`key` is already a CK_OBJECT_HANDLE, no ?
So better just cast to (void *) to appease the compiler.